### ISSUE TITLE
If callable A's params are assignable to callable B's params, A is not necessarily assignable to B

### ISSUE THREAD

**Issue by thepetertessier** [astral-sh/ty#2901]
### Question

Here is a simple example ([playground](https://play.ty.dev/20e04e38-76f5-43f9-993c-2b1926340fd7)):

```python
from collections.abc import Callable, Sequence

def higher_order(func: Callable[[Sequence], None]):
    pass

def my_func(s: list) -> None:
    pass

higher_order(my_func)  # invalid-argument-type
```

Here, `list` is assignable to `Sequence`, but `my_func` is not assignable to `func`.

Is this expected behavior? Am I missing something? Apologies if this is a known issue, wasn't able to find it!

### Version

ty 0.0.18

**Comment by AlexWaygood** [comment]
Hi! This is expected behaviour due to contravariance of parameters. `higher_order` states that it can only be passed a function that accepts _any_ `Sequence`. But `my_func` does not satisfy that requirement: it only accepts sequences that are instances of `list`.

**Comment by carljm** [comment]
To make this more intuitive, consider that the body of `higher_order` could include code that calls `func` with a dictionary, or a tuple, and that code would type check. So it wouldn't be safe to pass `higher_order` a function that can't handle those calls.

It would be fine the other way around -- if `higher_order` expects `Callable[[list], None]`, you can pass it a callable that takes any Sequence: https://play.ty.dev/268e1beb-1139-490b-97c1-7dbd89569f30

**Comment by thepetertessier** [comment]
Wow I'm blown away by the fast and well-explained response! A bit counterintuitive at first, but that makes a lot of sense!
